core/mutator/entry/mutator.go

```diff
@@ -61,6 +63,67 @@ func IsValidEntry(signedEntry *pb.SignedEntry) error {
 	return verifyKeys(ks, signedEntry.GetEntry(), signedEntry.GetSignatures())
 }
 
+// ReduceFn decides which of multiple updates can be applied in this revision.
+func ReduceFn(index []byte, msgs []*pb.EntryUpdate, leaves []*tpb.MapLeaf,
+	emit func(*tpb.MapLeaf), emitErr func(error)) {
+	if got := len(leaves); got > 1 {
+		emitErr(fmt.Errorf("expected 0 or 1 map leaf for index %x, got %v", index, got))
+		return // A bad index should not cause the whole batch to fail.
+	}
+	var oldValue *pb.SignedEntry // If no map leaf was found, oldValue will be nil.
+	var err error
+	if len(leaves) > 0 {
+		oldValue, err = FromLeafValue(leaves[0].GetLeafValue())
+		if err != nil {
+			emitErr(fmt.Errorf("entry.FromLeafValue(): %v", err))
+			return // A bad index should not cause the whole batch to fail.
+		}
+	}
+
+	if len(msgs) == 0 {
+		emitErr(fmt.Errorf("no msgs for index %x", index))
+		return // A bad index should not cause the whole batch to fail.
+	}
+
+	// Filter for mutations that are valid.
+	newEntries := make([]*pb.EntryUpdate, 0, len(msgs))
+	for i, msg := range msgs {
+		newValue, err := MutateFn(oldValue, msg.Mutation)
+		if err != nil {
+			emitErr(fmt.Errorf("entry: ReduceFn(%x, msg %d/%d): %v", index, i, len(msgs)-1, err))
+			continue
+		}
+		newEntries = append(newEntries, &pb.EntryUpdate{
+			Mutation:  newValue,
+			Committed: msg.Committed,
+		})
+	}
+	if len(newEntries) == 0 {
+		emitErr(fmt.Errorf("entry: no valid mutations for index %x", index))
+		return // No valid mutations for one index not cause the whole batch to fail.
+	}
+	// Choose the mutation deterministically, regardless of the messages order.
+	sort.Slice(newEntries, func(i, j int) bool {
```

Comment:
Minor nit: this is slightly wasteful to sort the whole slice and then pick the smallest element. This op can be performed in O(n). We don't expect large n though, so it's fine.